### PR TITLE
OCT-1652: Implement max MR funding cap

### PR DESCRIPTION
**Pull request by kgarbacinski**
## Definition of Done

1. [ ] Acceptance criteria are met.
2. [ ] PR is manually tested before the merge by developer(s).
    - [ ] Happy path is manually checked.
3. [ ] PR is manually tested by QA when their assistance is required (1).
    - [ ] Octant Areas & Test Cases are checked for impact and updated if required (2).
4. [ ] Unit tests are added unless there is a reason to omit them.
5. [ ] Automated tests are added when required.
6. [ ] The code is merged.
7. [ ] Tech documentation is added / updated, reviewed and approved (including mandatory approval by a code owner, should such exist for changed files).
    - [ ] BE: Swagger documentation is updated.
8. [ ] When required by QA:
    - [ ] Deployed to the relevant environment.
    - [ ] Passed system tests.

---

(1) Developer(s) in coordination with QA decide whether it's required. For small tickets introducing small changes QA assistance is most probably not required.

(2) [Octant Areas & Test Cases](https://docs.google.com/spreadsheets/d/1cRe6dxuKJV3a4ZskAwWEPvrFkQm6rEfyUCYwLTYw_Cc).

### CODE REVIEW

File: backend/tests/engine/projects/conftest.py [hunk 1]
```diff
@@ -77,3 +81,76 @@ def data_for_quadratic_funding(projects):
...
+        ProjectSumAllocationsDTO(projects[3], 4000),
+        ProjectSumAllocationsDTO(projects[4], 5000),
+        ProjectSumAllocationsDTO(projects[5], 6000),
+        ProjectSumAllocationsDTO(projects[6], 7000),
+        ProjectSumAllocationsDTO(projects[7], 8000),
+        ProjectSumAllocationsDTO(projects[8], 9000),
+        ProjectSumAllocationsDTO(projects[9], 10000),
+    ]
+
+    expected_distribution = {
```
Comment (leoni-q) [inline backend/tests/engine/projects/conftest.py:103]: It seems incorrect to me, `matched_rewards` is 35000, but sum of capped matched rewards is 55000

File: backend/app/engine/projects/rewards/allocations/__init__.py [hunk 2]
```diff
@@ -11,6 +14,11 @@ class AllocationItem:
     amount: int
 
 
+@dataclass(frozen=True)
+class ProjectSumAllocationsDTO(AllocationItem, JSONWizard):
+    amount: Union[Decimal, int]
```
[the remark this answers is not in the document]
Comment (kgarbacinski) [reply backend/app/engine/projects/rewards/allocations/__init__.py:19]: ProjectSumAllocationsDTO can represent the result of calculating the sum aligned with QF mechanism that returns Decimals, not ints (it sums roots of user's allocations). We operate on wei in terms of ```AllocationItem```, the sum can vary.

File: backend/tests/engine/projects/conftest.py (hunk 1 above)
Comment (kgarbacinski) [reply backend/tests/engine/projects/conftest.py:103]: changed

File: backend/app/engine/projects/rewards/funding_cap/percent.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                capped_quadratic_distribution < cap_amount
+            ]
+            uncapped_distribution = (
+                uncapped_projects / Decimal(uncapped_projects.sum()) * surplus
+            ).apply(Decimal)
+
+            new_surplus = self._calculate_surplus_to_project(
+                cap_amount, capped_quadratic_distribution, uncapped_distribution
+            )
+            surplus = new_surplus
```
Comment (0xartem) [inline backend/app/engine/projects/rewards/funding_cap/percent.py:70]: I can't figure out what this code is doing? Why do we need surplus here? I there any doc for that?

**Review by 0xartem — CHANGES_REQUESTED** [review]

File: backend/app/engine/projects/rewards/funding_cap/percent.py (hunk 3 above)
Comment (kgarbacinski) [reply backend/app/engine/projects/rewards/funding_cap/percent.py:70]: It's based on the .zip attached here: https://linear.app/golemfoundation/issue/OCT-1652/implement-max-mr-funding-cap-server. It was delivered by @andrzej-regulski.

Comment (kgarbacinski) [reply backend/app/engine/projects/rewards/funding_cap/percent.py:70]: AFAIK the surplus is basically for spreading the values over 0.2 of matched rewards for other projects that haven't reached the threshold yet

File: backend/tests/engine/projects/conftest.py [hunk 4]
```diff
@@ -77,3 +83,98 @@ def data_for_quadratic_funding(projects):
...
+        AllocationItem(projects[5], 6000),
+    ]
+
+    return matched_rewards, allocations
+
+
+@pytest.fixture
+def matched_rewards_with_capped_distribution(projects):
+    matched_rewards = 10000
+    computed_matched_rewards = [
```
[the remark this answers is not in the document]
Comment (kgarbacinski) [reply backend/tests/engine/projects/conftest.py:146]: https://chat.wildland.dev/wildland/pl/u4tnbi6ndbdj5e9khzn5orfpko

[the remark this answers is not in the document]
Comment (kgarbacinski) [reply backend/tests/engine/projects/conftest.py:146]: @leoni-q i've created a task for the corner case: https://linear.app/golemfoundation/issue/OCT-1682/handle-leftovers-from-capped-mr 

i'd leave the values you suggested as it is since that's a minor thing and we have other hurrying us up things

**Review by leoni-q — APPROVED** [review]
